File: src/main/java/net/coderextreme/data/arcold.java
package net.coderextreme.data;
import org.web3d.x3d.jsail.*;
import org.web3d.x3d.jsail.CADGeometry.*;
import org.web3d.x3d.jsail.Core.*;
import org.web3d.x3d.jsail.CubeMapTexturing.*;
import org.web3d.x3d.jsail.DIS.*;
import org.web3d.x3d.jsail.EnvironmentalEffects.*;
import org.web3d.x3d.jsail.EnvironmentalSensor.*;
import org.web3d.x3d.jsail.EventUtilities.*;
import org.web3d.x3d.jsail.Followers.*;
import org.web3d.x3d.jsail.Geometry2D.*;
import org.web3d.x3d.jsail.Geometry3D.*;
import org.web3d.x3d.jsail.Geospatial.*;
import org.web3d.x3d.jsail.Grouping.*;
import org.web3d.x3d.jsail.HAnim.*;
import org.web3d.x3d.jsail.Interpolation.*;
import org.web3d.x3d.jsail.KeyDeviceSensor.*;
import org.web3d.x3d.jsail.Layering.*;
import org.web3d.x3d.jsail.Layout.*;
import org.web3d.x3d.jsail.Lighting.*;
import org.web3d.x3d.jsail.NURBS.*;
import org.web3d.x3d.jsail.Navigation.*;
import org.web3d.x3d.jsail.Networking.*;
import org.web3d.x3d.jsail.ParticleSystems.*;
import org.web3d.x3d.jsail.Picking.*;
import org.web3d.x3d.jsail.PointingDeviceSensor.*;
import org.web3d.x3d.jsail.Rendering.*;
import org.web3d.x3d.jsail.RigidBodyPhysics.*;
import org.web3d.x3d.jsail.Scripting.*;
import org.web3d.x3d.jsail.Shaders.*;
import org.web3d.x3d.jsail.Shape.*;
import org.web3d.x3d.jsail.Sound.*;
import org.web3d.x3d.jsail.Text.*;
import org.web3d.x3d.jsail.Texturing3D.*;
import org.web3d.x3d.jsail.Texturing.*;
import org.web3d.x3d.jsail.Time.*;
import org.web3d.x3d.jsail.VolumeRendering.*;
import org.web3d.x3d.jsail.fields.*;
import java.util.ArrayList;
import java.util.List;
import net.coderextreme.X3DRoots;
public class arcold implements X3DRoots {
  public static void main(String[] args) {
    ConfigurationProperties.setXsltEngine(ConfigurationProperties.XSLT_ENGINE_NATIVE_JAVA);
    ConfigurationProperties.setDeleteIntermediateFiles(false);
    ConfigurationProperties.setStripTrailingZeroes(true);
    ConfigurationProperties.setStripDefaultAttributes(true);
    X3D model = new arcold().getRootNodeList().get(0); // only get one root node
    System.out.print(model.validationReport().trim());
    model.toFileX3D("../data/arcold.new.java.x3d");
    model.toFileJSON("../data/arcold.new.java.x3dj");
    }
    public List<X3D> getRootNodeList() {
    	List<X3D> list = new ArrayList<X3D>(1);
    	list.add(initialize());
    	return list;
    }
    public X3D initialize() {
ProtoInstance ProtoInstance0 = null;
ProtoInstance ProtoInstance1 = null;
ProtoInstance ProtoInstance2 = null;
ProtoInstance ProtoInstance3 = null;
ProtoInstance ProtoInstance4 = null;
ProtoInstance ProtoInstance5 = null;
ProtoInstance ProtoInstance6 = null;
ProtoInstance ProtoInstance7 = null;
ProtoInstance ProtoInstance8 = null;
ProtoInstance ProtoInstance9 = null;
ProtoInstance ProtoInstance10 = null;
ProtoInstance ProtoInstance11 = null;
ProtoInstance ProtoInstance12 = null;
      X3D X3D0 =  new X3D().setProfile(new SFString("Immersive")).setVersion(new SFString("4.0"))
      .setHead(new head()
        .addMeta(new meta().setName(new SFString("title")).setContent(new SFString("arcold.x3d")))
        .addMeta(new meta().setName(new SFString("creator")).setContent(new SFString("John Carlson")))
        .addMeta(new meta().setName(new SFString("generator")).setContent(new SFString("manual")))
        .addMeta(new meta().setName(new SFString("identifier")).setContent(new SFString("https://coderextreme.net/X3DJSONLD/src/main/data/arc.x3d")))
        .addMeta(new meta().setName(new SFString("description")).setContent(new SFString("an attempt to implement an arc in a graph"))))
      .setScene(new Scene()
        .addChild(new Viewpoint().setPosition(new float[] {0f ,0f ,5f }).setDescription(new SFString("a moving graph")))
        .addChild(new Background().setSkyColor(new MFColor0().getArray()))
        .addChild(new Transform().setDEF(new SFString("trans1"))
          .addChild(new Transform().setDEF(new SFString("rotscale1"))
            .addChild(new Shape()
              .setAppearance(new Appearance()
                .setMaterial(new Material().setDiffuseColor(new float[] {0.2f ,0.7f ,0.7f })))
              .setGeometry(new Cylinder().setRadius(0.1f )))))
        .addChild(new Transform().setDEF(new SFString("trans2"))
          .addChild(new Transform().setDEF(new SFString("rotscale2"))
            .addChild(new Shape()
              .setAppearance(new Appearance()
                .setMaterial(new Material().setDiffuseColor(new float[] {0.2f ,0.7f ,0.7f })))
              .setGeometry(new Cylinder().setRadius(0.1f )))))
        .addChild(new Transform().setDEF(new SFString("trans3"))
          .addChild(new Transform().setDEF(new SFString("rotscale3"))
            .addChild(new Shape()
              .setAppearance(new Appearance()
                .setMaterial(new Material().setDiffuseColor(new float[] {0.2f ,0.7f ,0.7f })))
              .setGeometry(new Cylinder().setRadius(0.1f )))))
        .addChild(new ProtoDeclare().setName(new SFString("point"))
          .setProtoInterface(new ProtoInterface()
            .addField(new field().setType("SFVec3f").setName(new SFString("translation")).setAccessType(field.ACCESSTYPE_INPUTOUTPUT).setValue(new SFString("0 0 0"))))
          .setProtoBody(new ProtoBody()
            .addChild(new Transform().setDEF(new SFString("node"))
              .setIS(new IS()
                .addConnect(new connect().setNodeField(new SFString("translation")).setProtoField(new SFString("translation"))))
              .addChild(new Shape()
                .setGeometry(new Sphere().setRadius(0.1f ))
                .setAppearance(new Appearance()
                  .setMaterial(new Material().setDiffuseColor(new float[] {1f ,0f ,0f }))))
              .addChild(new PositionInterpolator().setDEF(new SFString("PI1")).setKey(new MFFloat1().getArray()).setKeyValue(new MFVec3f2().getArray()))
              .addChild(new Script().setDEF(new SFString("MB1"))
                .addField(new field().setType("SFVec3f").setName(new SFString("translation")).setAccessType(field.ACCESSTYPE_INPUTOUTPUT).setValue(new SFString("50 50 0")))
                .addField(new field().setType("SFVec3f").setName(new SFString("old")).setAccessType(field.ACCESSTYPE_INPUTOUTPUT).setValue(new SFString("0 0 0")))
                .addField(new field().setType("SFTime").setName(new SFString("set_location")).setAccessType(field.ACCESSTYPE_INPUTONLY))
                .addField(new field().setType("MFVec3f").setName(new SFString("keyValue")).setAccessType(field.ACCESSTYPE_OUTPUTONLY))
                .setSourceCode("ecmascript:\n"+
"		function set_location(value) {\n"+
"                    old = translation;\n"+
"		    translation = new SFVec3f(Math.random()*10-5, Math.random()*10-5, Math.random()*10-5);\n"+
"                    keyValue = new MFVec3f([old, translation]);\n"+
"		    // Browser.println(translation);\n"+
"		}"))
              .addChild(new TimeSensor().setDEF(new SFString("CL1")).setCycleInterval(3d).setLoop(true))
              .addChild(new ROUTE().setFromNode(new SFString("CL1")).setFromField(new SFString("cycleTime")).setToNode(new SFString("MB1")).setToField(new SFString("set_location")))
              .addChild(new ROUTE().setFromNode(new SFString("CL1")).setFromField(new SFString("fraction_changed")).setToNode(new SFString("PI1")).setToField(new SFString("set_fraction")))
              .addChild(new ROUTE().setFromNode(new SFString("MB1")).setFromField(new SFString("keyValue")).setToNode(new SFString("PI1")).setToField(new SFString("keyValue")))
              .addChild(new ROUTE().setFromNode(new SFString("PI1")).setFromField(new SFString("value_changed")).setToNode(new SFString("node")).setToField(new SFString("set_translation"))))))
        .addComments(new CommentsBlock("from doug sanden"))
        .addChild(new ProtoDeclare().setName(new SFString("x3dconnector"))
          .setProtoInterface(new ProtoInterface()
            .addField(new field().setType("SFNode").setName(new SFString("startnode")).setAccessType(field.ACCESSTYPE_INPUTOUTPUT))
            .addField(new field().setType("SFNode").setName(new SFString("endnode")).setAccessType(field.ACCESSTYPE_INPUTOUTPUT))
            .addField(new field().setType("SFNode").setName(new SFString("transnode")).setAccessType(field.ACCESSTYPE_INPUTOUTPUT))
            .addField(new field().setType("SFNode").setName(new SFString("rotscalenode")).setAccessType(field.ACCESSTYPE_INPUTOUTPUT))
            .addField(new field().setType("SFVec3f").setName(new SFString("set_startpoint")).setAccessType(field.ACCESSTYPE_INPUTONLY))
            .addField(new field().setType("SFVec3f").setName(new SFString("set_endpoint")).setAccessType(field.ACCESSTYPE_INPUTONLY)))
          .setProtoBody(new ProtoBody()
            .addChild(new Script().setDEF(new SFString("S1"))
              .addField(new field().setType("SFNode").setName(new SFString("startnode")).setAccessType(field.ACCESSTYPE_INPUTOUTPUT))
              .addField(new field().setType("SFNode").setName(new SFString("endnode")).setAccessType(field.ACCESSTYPE_INPUTOUTPUT))
              .addField(new field().setType("SFNode").setName(new SFString("transnode")).setAccessType(field.ACCESSTYPE_INPUTOUTPUT))
              .addField(new field().setType("SFNode").setName(new SFString("rotscalenode")).setAccessType(field.ACCESSTYPE_INPUTOUTPUT))
              .addField(new field().setType("SFVec3f").setName(new SFString("set_startpoint")).setAccessType(field.ACCESSTYPE_INPUTONLY))
              .addField(new field().setType("SFVec3f").setName(new SFString("set_endpoint")).setAccessType(field.ACCESSTYPE_INPUTONLY))
              .setIS(new IS()
                .addConnect(new connect().setNodeField(new SFString("startnode")).setProtoField(new SFString("startnode")))
                .addConnect(new connect().setNodeField(new SFString("endnode")).setProtoField(new SFString("endnode")))
                .addConnect(new connect().setNodeField(new SFString("transnode")).setProtoField(new SFString("transnode")))
                .addConnect(new connect().setNodeField(new SFString("rotscalenode")).setProtoField(new SFString("rotscalenode")))
                .addConnect(new connect().setNodeField(new SFString("set_startpoint")).setProtoField(new SFString("set_startpoint")))
                .addConnect(new connect().setNodeField(new SFString("set_endpoint")).setProtoField(new SFString("set_endpoint"))))
              .setSourceCode("ecmascript:\n"+
"        function recompute(startpoint,endpoint){\n"+
"	    if (typeof endpoint === 'undefined') {\n"+
"		return;\n"+
"	    }\n"+
"            var dif = endpoint.subtract(startpoint);\n"+
"            var dist = dif.length()*0.5;\n"+
"            var dif2 = dif.multiply(0.5);\n"+
"            var norm = dif.normalize();\n"+
"            var transl = startpoint.add(dif2);\n"+
"	    if (typeof Quaternion !== 'undefined') {\n"+
"		    return {\n"+
"			    scale : new SFVec3f(1.0,dist,1.0),\n"+
"			    translation : transl,\n"+
"			    rotation : new Quaternion.rotateFromTo(new SFVec3f(0.0,1.0,0.0), norm)\n"+
"		    };\n"+
"	    } else {\n"+
"		    return {\n"+
"			    scale : new SFVec3f(1.0,dist,1.0),\n"+
"			    translation : transl,\n"+
"			    rotation : new SFRotation(new SFVec3f(0.0,1.0,0.0),norm)\n"+
"		    };\n"+
"	    }\n"+
"	}\n"+
"	function recompute_and_route(startpoint, endpoint) {\n"+
"		var trafo = recompute(startpoint, endpoint);\n"+
"		if (typeof trafo !== 'undefined') {\n"+
"			transnode.translation = trafo.translation;\n"+
"			rotscalenode.rotation = trafo.rotation;\n"+
"			rotscalenode.scale = trafo.scale;\n"+
"		} else {\n"+
"			Browser.print(\"recompute returned undefined\");\n"+
"		}\n"+
"	}\n"+
"        function initialize(){\n"+
"            recompute_and_route(startnode.translation,endnode.translation);\n"+
"        }\n"+
"        function set_startpoint(val,t){\n"+
"            recompute_and_route(val || startnode.translation,endnode.translation);\n"+
"        }\n"+
"        function set_endpoint(val,t){\n"+
"            recompute_and_route(startnode.translation,val || endnode.translation);\n"+
"        }"))))
        .addChild(ProtoInstance0 = new ProtoInstance().setName(new SFString("point")).setDEF(new SFString("G1")))
        .addChild(ProtoInstance1 = new ProtoInstance().setName(new SFString("point")).setDEF(new SFString("G2")))
        .addChild(ProtoInstance2 = new ProtoInstance().setName(new SFString("point")).setDEF(new SFString("G3")))
        .addChild(ProtoInstance3 = new ProtoInstance().setName(new SFString("point")).setDEF(new SFString("G4")))
        .addChild(ProtoInstance4 = new ProtoInstance().setName(new SFString("x3dconnector")).setDEF(new SFString("connector1")))
        .addChild(ProtoInstance7 = new ProtoInstance().setName(new SFString("x3dconnector")).setDEF(new SFString("connector2")))
        .addChild(ProtoInstance10 = new ProtoInstance().setName(new SFString("x3dconnector")).setDEF(new SFString("connector3")))
        .addChild(new ROUTE().setFromNode(new SFString("G1")).setFromField(new SFString("translation_changed")).setToNode(new SFString("connector1")).setToField(new SFString("set_startpoint")))
        .addChild(new ROUTE().setFromNode(new SFString("G2")).setFromField(new SFString("translation_changed")).setToNode(new SFString("connector1")).setToField(new SFString("set_endpoint")))
        .addChild(new ROUTE().setFromNode(new SFString("G1")).setFromField(new SFString("translation_changed")).setToNode(new SFString("connector2")).setToField(new SFString("set_startpoint")))
        .addChild(new ROUTE().setFromNode(new SFString("G3")).setFromField(new SFString("translation_changed")).setToNode(new SFString("connector2")).setToField(new SFString("set_endpoint")))
        .addChild(new ROUTE().setFromNode(new SFString("G1")).setFromField(new SFString("translation_changed")).setToNode(new SFString("connector3")).setToField(new SFString("set_startpoint")))
        .addChild(new ROUTE().setFromNode(new SFString("G4")).setFromField(new SFString("translation_changed")).setToNode(new SFString("connector3")).setToField(new SFString("set_endpoint"))));
ProtoInstance4
          .addFieldValue(new fieldValue().setName(new SFString("startnode"))
            .addChild(ProtoInstance5 = new ProtoInstance().setName(new SFString("point")).setUSE(new SFString("G1"))));
ProtoInstance4
          .addFieldValue(new fieldValue().setName(new SFString("endnode"))
            .addChild(ProtoInstance6 = new ProtoInstance().setName(new SFString("point")).setUSE(new SFString("G2"))));
ProtoInstance4
          .addFieldValue(new fieldValue().setName(new SFString("transnode"))
            .addChild(new Transform().setUSE(new SFString("trans1"))));
ProtoInstance4
          .addFieldValue(new fieldValue().setName(new SFString("rotscalenode"))
            .addChild(new Transform().setUSE(new SFString("rotscale1"))));
ProtoInstance7
          .addFieldValue(new fieldValue().setName(new SFString("startnode"))
            .addChild(ProtoInstance8 = new ProtoInstance().setName(new SFString("point")).setUSE(new SFString("G1"))));
ProtoInstance7
          .addFieldValue(new fieldValue().setName(new SFString("endnode"))
            .addChild(ProtoInstance9 = new ProtoInstance().setName(new SFString("point")).setUSE(new SFString("G3"))));
ProtoInstance7
          .addFieldValue(new fieldValue().setName(new SFString("transnode"))
            .addChild(new Transform().setUSE(new SFString("trans2"))));
ProtoInstance7
          .addFieldValue(new fieldValue().setName(new SFString("rotscalenode"))
            .addChild(new Transform().setUSE(new SFString("rotscale2"))));
ProtoInstance10
          .addFieldValue(new fieldValue().setName(new SFString("startnode"))
            .addChild(ProtoInstance11 = new ProtoInstance().setName(new SFString("point")).setUSE(new SFString("G1"))));
ProtoInstance10
          .addFieldValue(new fieldValue().setName(new SFString("endnode"))
            .addChild(ProtoInstance12 = new ProtoInstance().setName(new SFString("point")).setUSE(new SFString("G4"))));
ProtoInstance10
          .addFieldValue(new fieldValue().setName(new SFString("transnode"))
            .addChild(new Transform().setUSE(new SFString("trans3"))));
ProtoInstance10
          .addFieldValue(new fieldValue().setName(new SFString("rotscalenode"))
            .addChild(new Transform().setUSE(new SFString("rotscale3"))));
    return X3D0;
    }
private class MFColor0 {
  private org.web3d.x3d.jsail.fields.MFColor getArray() {
    return new org.web3d.x3d.jsail.fields.MFColor(new float[] {0.4f ,0.4f ,0.4f });
  }
}
private class MFFloat1 {
  private org.web3d.x3d.jsail.fields.MFFloat getArray() {
    return new org.web3d.x3d.jsail.fields.MFFloat(new float[] {0f ,1f });
  }
}
private class MFVec3f2 {
  private org.web3d.x3d.jsail.fields.MFVec3f getArray() {
    return new org.web3d.x3d.jsail.fields.MFVec3f(new float[] {0f ,0f ,0f ,0f ,5f ,0f });
  }
}
}

File: src/main/node/net/coderextreme/data/arcold.js
var java = require('java');
var autoclass = require('../../../X3Dautoclass.js');
var ConfigurationProperties = autoclass.ConfigurationProperties;
ConfigurationProperties.showDefaultAttributes = false;
ConfigurationProperties.xsltEngine = ConfigurationProperties.XSLT_ENGINE_NATIVE_JAVA;
ConfigurationProperties.deleteIntermediateFiles = false;
ConfigurationProperties.setStripTrailingZeroesSync(true);
var ProtoInstance0 = null;
var ProtoInstance1 = null;
var ProtoInstance2 = null;
var ProtoInstance3 = null;
var ProtoInstance4 = null;
var ProtoInstance5 = null;
var ProtoInstance6 = null;
var ProtoInstance7 = null;
var ProtoInstance8 = null;
var ProtoInstance9 = null;
var ProtoInstance10 = null;
var ProtoInstance11 = null;
var ProtoInstance12 = null;
      var X3D0 =  new autoclass.X3D().setProfileSync("Immersive").setVersionSync("3.3")
      .setHeadSync(new autoclass.head()
        .addMetaSync(new autoclass.meta().setNameSync("title").setContentSync("arcold.x3d"))
        .addMetaSync(new autoclass.meta().setNameSync("creator").setContentSync("John Carlson"))
        .addMetaSync(new autoclass.meta().setNameSync("generator").setContentSync("manual"))
        .addMetaSync(new autoclass.meta().setNameSync("identifier").setContentSync("https://coderextreme.net/X3DJSONLD/arc.x3d"))
        .addMetaSync(new autoclass.meta().setNameSync("description").setContentSync("an attempt to implement an arc in a graph")))
      .setSceneSync(new autoclass.Scene()
        .addChildSync(new autoclass.Viewpoint().setPositionSync(java.newArray("float", [java.newFloat(0), java.newFloat(0), java.newFloat(5)])).setDescriptionSync("a moving graph"))
        .addChildSync(new autoclass.Background().setSkyColorSync(java.newArray("float", [java.newFloat(0.4), java.newFloat(0.4), java.newFloat(0.4)])))
        .addChildSync(new autoclass.Transform().setDEFSync("trans1")
          .addChildSync(new autoclass.Transform().setDEFSync("rotscale1")
            .addChildSync(new autoclass.Shape()
              .setAppearanceSync(new autoclass.Appearance()
                .setMaterialSync(new autoclass.Material().setDiffuseColorSync(java.newArray("float", [java.newFloat(0.2), java.newFloat(0.7), java.newFloat(0.7)]))))
              .setGeometrySync(new autoclass.Cylinder().setRadiusSync(java.newFloat(0.1))))))
        .addChildSync(new autoclass.Transform().setDEFSync("trans2")
          .addChildSync(new autoclass.Transform().setDEFSync("rotscale2")
            .addChildSync(new autoclass.Shape()
              .setAppearanceSync(new autoclass.Appearance()
                .setMaterialSync(new autoclass.Material().setDiffuseColorSync(java.newArray("float", [java.newFloat(0.2), java.newFloat(0.7), java.newFloat(0.7)]))))
              .setGeometrySync(new autoclass.Cylinder().setRadiusSync(java.newFloat(0.1))))))
        .addChildSync(new autoclass.Transform().setDEFSync("trans3")
          .addChildSync(new autoclass.Transform().setDEFSync("rotscale3")
            .addChildSync(new autoclass.Shape()
              .setAppearanceSync(new autoclass.Appearance()
                .setMaterialSync(new autoclass.Material().setDiffuseColorSync(java.newArray("float", [java.newFloat(0.2), java.newFloat(0.7), java.newFloat(0.7)]))))
              .setGeometrySync(new autoclass.Cylinder().setRadiusSync(java.newFloat(0.1))))))
        .addChildSync(new autoclass.ProtoDeclare().setNameSync("point")
          .setProtoInterfaceSync(new autoclass.ProtoInterface()
            .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFVEC3F).setNameSync("translation").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTOUTPUT).setValueSync("0 0 0")))
          .setProtoBodySync(new autoclass.ProtoBody()
            .addChildSync(new autoclass.Transform().setDEFSync("node")
              .setISSync(new autoclass.IS()
                .addConnectSync(new autoclass.connect().setNodeFieldSync("translation").setProtoFieldSync("translation")))
              .addChildSync(new autoclass.Shape()
                .setGeometrySync(new autoclass.Sphere().setRadiusSync(java.newFloat(0.1)))
                .setAppearanceSync(new autoclass.Appearance()
                  .setMaterialSync(new autoclass.Material().setDiffuseColorSync(java.newArray("float", [java.newFloat(1), java.newFloat(0), java.newFloat(0)])))))
              .addChildSync(new autoclass.PositionInterpolator().setDEFSync("PI1").setKeySync(java.newArray("float", [java.newFloat(0), java.newFloat(1)])).setKeyValueSync(java.newArray("float", [java.newFloat(0), java.newFloat(0), java.newFloat(0), java.newFloat(0), java.newFloat(5), java.newFloat(0)])))
              .addChildSync(new autoclass.Script().setDEFSync("MB1")
                .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFVEC3F).setNameSync("translation").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTOUTPUT).setValueSync("50 50 0"))
                .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFVEC3F).setNameSync("old").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTOUTPUT).setValueSync("0 0 0"))
                .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFTIME).setNameSync("set_location").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTONLY))
                .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_MFVEC3F).setNameSync("keyValue").setAccessTypeSync(autoclass.field.ACCESSTYPE_OUTPUTONLY))
                .setSourceCodeSync("ecmascript:\n"+
"		function set_location(value) {\n"+
"                    old = translation;\n"+
"		    translation = new SFVec3f(Math.random()*10-5, Math.random()*10-5, Math.random()*10-5);\n"+
"                    keyValue = new MFVec3f([old, translation]);\n"+
"		    // Browser.println(translation);\n"+
"		}"))
              .addChildSync(new autoclass.TimeSensor().setDEFSync("CL1").setCycleIntervalSync(3).setLoopSync(true))
              .addChildSync(new autoclass.ROUTE().setFromNodeSync("CL1").setFromFieldSync("cycleTime").setToNodeSync("MB1").setToFieldSync("set_location"))
              .addChildSync(new autoclass.ROUTE().setFromNodeSync("CL1").setFromFieldSync("fraction_changed").setToNodeSync("PI1").setToFieldSync("set_fraction"))
              .addChildSync(new autoclass.ROUTE().setFromNodeSync("MB1").setFromFieldSync("keyValue").setToNodeSync("PI1").setToFieldSync("keyValue"))
              .addChildSync(new autoclass.ROUTE().setFromNodeSync("PI1").setFromFieldSync("value_changed").setToNodeSync("node").setToFieldSync("set_translation")))))
        .addCommentsSync(new autoclass.CommentsBlock("from doug sanden"))
        .addChildSync(new autoclass.ProtoDeclare().setNameSync("x3dconnector")
          .setProtoInterfaceSync(new autoclass.ProtoInterface()
            .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFNODE).setNameSync("startnode").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTOUTPUT))
            .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFNODE).setNameSync("endnode").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTOUTPUT))
            .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFNODE).setNameSync("transnode").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTOUTPUT))
            .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFNODE).setNameSync("rotscalenode").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTOUTPUT))
            .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFVEC3F).setNameSync("set_startpoint").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTONLY))
            .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFVEC3F).setNameSync("set_endpoint").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTONLY)))
          .setProtoBodySync(new autoclass.ProtoBody()
            .addChildSync(new autoclass.Script().setDEFSync("S1")
              .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFNODE).setNameSync("startnode").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTOUTPUT))
              .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFNODE).setNameSync("endnode").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTOUTPUT))
              .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFNODE).setNameSync("transnode").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTOUTPUT))
              .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFNODE).setNameSync("rotscalenode").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTOUTPUT))
              .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFVEC3F).setNameSync("set_startpoint").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTONLY))
              .addFieldSync(new autoclass.field().setTypeSync(autoclass.field.TYPE_SFVEC3F).setNameSync("set_endpoint").setAccessTypeSync(autoclass.field.ACCESSTYPE_INPUTONLY))
              .setISSync(new autoclass.IS()
                .addConnectSync(new autoclass.connect().setNodeFieldSync("startnode").setProtoFieldSync("startnode"))
                .addConnectSync(new autoclass.connect().setNodeFieldSync("endnode").setProtoFieldSync("endnode"))
                .addConnectSync(new autoclass.connect().setNodeFieldSync("transnode").setProtoFieldSync("transnode"))
                .addConnectSync(new autoclass.connect().setNodeFieldSync("rotscalenode").setProtoFieldSync("rotscalenode"))
                .addConnectSync(new autoclass.connect().setNodeFieldSync("set_startpoint").setProtoFieldSync("set_startpoint"))
                .addConnectSync(new autoclass.connect().setNodeFieldSync("set_endpoint").setProtoFieldSync("set_endpoint")))
              .setSourceCodeSync("ecmascript:\n"+
"        function recompute(startpoint,endpoint){\n"+
"	    if (typeof endpoint === 'undefined') {\n"+
"		return;\n"+
"	    }\n"+
"            var dif = endpoint.subtract(startpoint);\n"+
"            var dist = dif.length()*0.5;\n"+
"            var dif2 = dif.multiply(0.5);\n"+
"            var norm = dif.normalize();\n"+
"            var transl = startpoint.add(dif2);\n"+
"	    if (typeof Quaternion !== 'undefined') {\n"+
"		    return {\n"+
"			    scale : new SFVec3f(1.0,dist,1.0),\n"+
"			    translation : transl,\n"+
"			    rotation : new Quaternion.rotateFromTo(new SFVec3f(0.0,1.0,0.0), norm)\n"+
"		    };\n"+
"	    } else {\n"+
"		    return {\n"+
"			    scale : new SFVec3f(1.0,dist,1.0),\n"+
"			    translation : transl,\n"+
"			    rotation : new SFRotation(new SFVec3f(0.0,1.0,0.0),norm)\n"+
"		    };\n"+
"	    }\n"+
"	}\n"+
"	function recompute_and_route(startpoint, endpoint) {\n"+
"		var trafo = recompute(startpoint, endpoint);\n"+
"		if (typeof trafo !== 'undefined') {\n"+
"			transnode.translation = trafo.translation;\n"+
"			rotscalenode.rotation = trafo.rotation;\n"+
"			rotscalenode.scale = trafo.scale;\n"+
"		} else {\n"+
"			Browser.print(\"recompute returned undefined\");\n"+
"		}\n"+
"	}\n"+
"        function initialize(){\n"+
"            recompute_and_route(startnode.translation,endnode.translation);\n"+
"        }\n"+
"        function set_startpoint(val,t){\n"+
"            recompute_and_route(val || startnode.translation,endnode.translation);\n"+
"        }\n"+
"        function set_endpoint(val,t){\n"+
"            recompute_and_route(startnode.translation,val || endnode.translation);\n"+
"        }"))))
        .addChildSync(ProtoInstance0 = new autoclass.ProtoInstance().setNameSync("point").setDEFSync("G1"))
        .addChildSync(ProtoInstance1 = new autoclass.ProtoInstance().setNameSync("point").setDEFSync("G2"))
        .addChildSync(ProtoInstance2 = new autoclass.ProtoInstance().setNameSync("point").setDEFSync("G3"))
        .addChildSync(ProtoInstance3 = new autoclass.ProtoInstance().setNameSync("point").setDEFSync("G4"))
        .addChildSync(ProtoInstance4 = new autoclass.ProtoInstance().setNameSync("x3dconnector").setDEFSync("connector1"))
        .addChildSync(ProtoInstance7 = new autoclass.ProtoInstance().setNameSync("x3dconnector").setDEFSync("connector2"))
        .addChildSync(ProtoInstance10 = new autoclass.ProtoInstance().setNameSync("x3dconnector").setDEFSync("connector3"))
        .addChildSync(new autoclass.ROUTE().setFromNodeSync("G1").setFromFieldSync("translation_changed").setToNodeSync("connector1").setToFieldSync("set_startpoint"))
        .addChildSync(new autoclass.ROUTE().setFromNodeSync("G2").setFromFieldSync("translation_changed").setToNodeSync("connector1").setToFieldSync("set_endpoint"))
        .addChildSync(new autoclass.ROUTE().setFromNodeSync("G1").setFromFieldSync("translation_changed").setToNodeSync("connector2").setToFieldSync("set_startpoint"))
        .addChildSync(new autoclass.ROUTE().setFromNodeSync("G3").setFromFieldSync("translation_changed").setToNodeSync("connector2").setToFieldSync("set_endpoint"))
        .addChildSync(new autoclass.ROUTE().setFromNodeSync("G1").setFromFieldSync("translation_changed").setToNodeSync("connector3").setToFieldSync("set_startpoint"))
        .addChildSync(new autoclass.ROUTE().setFromNodeSync("G4").setFromFieldSync("translation_changed").setToNodeSync("connector3").setToFieldSync("set_endpoint")))      ;
ProtoInstance4
          .addFieldValueSync(new autoclass.fieldValue().setNameSync("startnode")
            .addChildSync(ProtoInstance5 = new autoclass.ProtoInstance().setUSESync("G1")));
ProtoInstance4
          .addFieldValueSync(new autoclass.fieldValue().setNameSync("endnode")
            .addChildSync(ProtoInstance6 = new autoclass.ProtoInstance().setUSESync("G2")));
ProtoInstance4
          .addFieldValueSync(new autoclass.fieldValue().setNameSync("transnode")
            .addChildSync(new autoclass.Transform().setUSESync("trans1")));
ProtoInstance4
          .addFieldValueSync(new autoclass.fieldValue().setNameSync("rotscalenode")
            .addChildSync(new autoclass.Transform().setUSESync("rotscale1")));
ProtoInstance7
          .addFieldValueSync(new autoclass.fieldValue().setNameSync("startnode")
            .addChildSync(ProtoInstance8 = new autoclass.ProtoInstance().setUSESync("G1")));
ProtoInstance7
          .addFieldValueSync(new autoclass.fieldValue().setNameSync("endnode")
            .addChildSync(ProtoInstance9 = new autoclass.ProtoInstance().setUSESync("G3")));
ProtoInstance7
          .addFieldValueSync(new autoclass.fieldValue().setNameSync("transnode")
            .addChildSync(new autoclass.Transform().setUSESync("trans2")));
ProtoInstance7
          .addFieldValueSync(new autoclass.fieldValue().setNameSync("rotscalenode")
            .addChildSync(new autoclass.Transform().setUSESync("rotscale2")));
ProtoInstance10
          .addFieldValueSync(new autoclass.fieldValue().setNameSync("startnode")
            .addChildSync(ProtoInstance11 = new autoclass.ProtoInstance().setUSESync("G1")));
ProtoInstance10
          .addFieldValueSync(new autoclass.fieldValue().setNameSync("endnode")
            .addChildSync(ProtoInstance12 = new autoclass.ProtoInstance().setUSESync("G4")));
ProtoInstance10
          .addFieldValueSync(new autoclass.fieldValue().setNameSync("transnode")
            .addChildSync(new autoclass.Transform().setUSESync("trans3")));
ProtoInstance10
          .addFieldValueSync(new autoclass.fieldValue().setNameSync("rotscalenode")
            .addChildSync(new autoclass.Transform().setUSESync("rotscale3")));
    X3D0.toFileX3D("../data/arcold.new.x3d");
